File: 백준/python/boj_14889.py
import sys

sys.stdin = open("input.txt", "r")

N = int(input())
board = [list(map(int, input().split(" "))) for _ in range(N)]

check = [False] * N
answer = 1e9
def dfs(idx, cnt):
    global  answer
    if cnt == N/2:
        start = 0
        link = 0
        for i in range(N):
            for j in range(N):
                if not check[i] and not check[j]:
                    start += board[i][j]
                if check[i] and check[j]:
                    link += board[i][j]
        answer = min(answer, abs(start-link))
        return
    for i in range(idx, N):
        if not check[i]:
            check[i] = True
            dfs(i+1, cnt+1)
            check[i] = False

dfs(0,0)
print(answer)

File: 백준/js/boj_14889.js
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
let input = require("fs").readFileSync(filePath).toString().trim().split("\n");

const N = Number(input.shift());
const board = input.map((v) => v.split(" ").map(Number));

const check = new Array(N).fill(false);
let answer = Number.MAX_VALUE;

function dfs(idx, cnt) {
  if (cnt === N / 2 && check[0]) {
    let start = 0;
    let link = 0;
    for (let i = 0; i < N; i++) {
      for (let j = 0; j < N; j++) {
        if (check[i] && check[j]) start += board[i][j];
        else if (!check[i] && !check[j]) link += board[i][j];
      }
    }
    answer = Math.min(answer, Math.abs(start - link));
    return;
  }

  for (let i = idx; i < N; i++) {
    if (!check[i]) {
      check[i] = true;
      dfs(i + 1, cnt + 1);
      check[i] = false;
    }
  }
}

dfs(0, 0);
console.log(answer);
